### PR TITLE
Move DataOutInterface::write_pvd/visit_record() to namespace DataOutBase.

### PR DESCRIPTION
**Pull request by bangerth**
These functions did not depend on the state of the DataOutInterface object and could therefore
be made 'static'. On the other hand, we have traditionally kept such functions in namespace
DataOutBase. Move them there, and deprecate the old versions of these functions.

### CODE REVIEW

**Comment by tjhei** [comment]
One disadvantage is that this separates ``write_pvtu_record`` (which does depend on the state) from ``write_pvd_record``. Any thoughts on this?

**Comment by bangerth** [comment]
Yes, I did think of making the function `static` in `DataOutInterface`.

What really alerted me to the problem in the first place is that we had code that looked like this:
```
  DataOut().write_pvd_record(...);
```
see https://github.com/geodynamics/aspect/blob/master/source/particle/output/vtu.cc#L230 . This seems to make little sense because if that works, it's obvious that `write_pvd_record()` does not depend on the *state* of the object. On the other hand, you can't write `DataOutInterface().write_pvd_record` because `DataOutInterface` is an abstract base class of which one cannot create objects. One may, however, have written `MatrixOut().write_pvd_record()`, which has exactly the same effect as `DataOut().write_pvd_record()` despite the fact that it is entirely obscure and unintuitive in most contexts why `MatrixOut` would enter the picture :-)

In summary, I felt that (i) using `DataOut().write_pvd_record()` is confusing, and (ii) `DataOutInterface().write_pvd_record()` would be intuitive but is no allowed. Consequently, I thought that removing all class association was the right thing to do.

I recognize that that rips similar functions apart. I would be ok to add links to each set of functions to the respective other set (i.e., links from `write_pvtu_record()` to the functions that are now in `DataOutBase`, and the other way around). Would that address your issues?

**Comment by tjhei** [comment]
> Would that address your issues?

I fully agree with your arguments: separating the functions is certainly preferable. 

One other idea I just had:
One could also make ``write_pvtu_record`` static and move it out, provided we ask the user to supply a list of variable names to the function. I am not sure that is worth it though.

Please merge yourself if you want to keep it as is.

**Comment by bangerth** [comment]
I moved the implementation of `write_pvtu_record` into the base class, with additional arguments. The existing function in `DataOutInterface` continues to exist, but simply calls the other function with the additional arguments.

Want to take another look?

File: doc/news/changes.h [hunk 1]
```diff
@@ -38,6 +38,18 @@ inconvenience this causes.
 </p>
 
 <ol>
+ <li> Deprecated: The DataOutInterface::write_pvd_record() and
+ DataOutInterface::write_visit_record() functions were actually
```
Comment (tjhei) [inline doc/news/changes.h]: also mention ``write_pvtu_record`` please.

File: include/deal.II/base/data_out_base.h [hunk 2]
```diff
@@ -1725,6 +1725,174 @@ namespace DataOutBase
...
+   *    std::ofstream output (filename.c_str());
+   *    data_out.write_vtu (output);
+   *
+   *    times_and_names.push_back (std::pair<double,std::string> (time, filename));
+   *    std::ofstream pvd_output ("solution.pvd");
+   *    DataOutBase::write_pvd_record (pvd_output, times_and_names);
+   *  }
+   * @endcode
+   *
+   * @note See DataOutInterface::write_vtu or DataOutInterface::write_pvtu_record
```
Comment (tjhei) [inline include/deal.II/base/data_out_base.h]: or ``DataOutInterface< dim, spacedim >::write_vtu_in_parallel``

File: include/deal.II/base/data_out_base.h [hunk 3]
```diff
@@ -2177,138 +2347,34 @@ class DataOutInterface
                           const std::vector<std::string> &piece_names) const;
```
Comment (tjhei) [inline include/deal.II/base/data_out_base.h]: should this be deprecated and the documentation be removed (just referencing the static version)?

File: doc/news/changes.h (hunk 1 above)
Comment (bangerth) [inline doc/news/changes.h]: Well, actually I didn't deprecate it, nor intend to. It's much easier to call the function in `DataOutInterface` than to call the one in `DataOutBase` because it knows where to get all of the information.

**Comment by bangerth** [comment]
So we may be miscommunicating here. I didn't intend to deprecate `DataOutInterface::write_pvtu_record()`, because it's substantially simpler to use than `DataOutBase::write_pvtu_record()`. As a consequence, I just extracted the basic functionality to the `DataOutBase` namespace, but left the original function in place (and consequently also didn't update any examples or testcases).

One could think of deprecating the old function if it was easy for users to get at the dataset names and vector-data-ranges of a `DataOutInterface` object. But it isn't -- these are both protected virtual member functions, so a user can't call them. I am loath to also expose these: we're rewritten this class a number of times, and these are internal interfaces that have changed over time. I suspect that they may change again if anyone ever comes around and implements output for tensor-valued objects, for example.

**Comment by bangerth** [comment]
I did address the one other comment you had, though.

File: include/deal.II/base/data_out_base.h [hunk 4]
```diff
@@ -1725,6 +1725,175 @@ namespace DataOutBase
                        const std::vector<std_cxx11::tuple<unsigned int, unsigned int, std::string> > &vector_data_ranges,
                        const VtkFlags                          &flags,
                        std::ostream                            &out);
+  /**
+   * Some visualization programs, such as ParaView, can read several separate
+   * VTU files to parallelize visualization. In that case, you need a
+   * <code>.pvtu</code> file that describes which VTU files (written, for
+   * example, through the DataOutInterface::write_vtu() function) form a group.
```
Comment (tjhei) [inline include/deal.II/base/data_out_base.h]: I think this description should have a big "You probably want to use ``DataOutInterface::write_pvtu_record`` rather than this function" note in the first paragraph.

File: include/deal.II/base/data_out_base.h [hunk 5]
```diff
@@ -2141,7 +2310,8 @@ class DataOutInterface
    * performance on parallel filesystems. Also see
    * DataOutInterface::write_vtu().
    */
-  void write_vtu_in_parallel (const char *filename, MPI_Comm comm) const;
+  void write_vtu_in_parallel (const char *filename,
+                              MPI_Comm comm) const;
 
   /**
    * Some visualization programs, such as ParaView, can read several separate
```
Comment (tjhei) [inline include/deal.II/base/data_out_base.h]: can you add something like ".. master record, by calling DataOutBase::write_pvtu_record" or something similar at the end of this paragraph?

**Comment by tjhei** [comment]
> So we may be miscommunicating here. 

Yes. My bad, I didn't check your changes closely enough.

> But it isn't -- these are both protected virtual member functions, so a user can't call them.

Argh, okay. That makes sense then. I am happy with it (see two small comments). Merge yourself when you are happy.
